DuckDuckGo/NewTabPage/PrivacyStats/NewTabPagePrivacyStatsModel.swift

```diff
@@ -0,0 +1,144 @@
+    private let trackerDataProvider: PrivacyStatsTrackerDataProviding
+
+    private let statsUpdateSubject = PassthroughSubject<Void, Never>()
+    private var cancellables: Set<AnyCancellable> = []
+
+    init(
+        privacyStats: PrivacyStatsCollecting,
+        trackerDataProvider: PrivacyStatsTrackerDataProviding,
+        settingsPersistor: NewTabPagePrivacyStatsSettingsPersistor = UserDefaultsNewTabPagePrivacyStatsSettingsPersistor()
+    ) {
+        self.privacyStats = privacyStats
+        self.trackerDataProvider = trackerDataProvider
+        self.settingsPersistor = settingsPersistor
+
+        isViewExpanded = settingsPersistor.isViewExpanded
+        statsUpdatePublisher = statsUpdateSubject.eraseToAnyPublisher()
+
+        privacyStats.statsUpdatePublisher
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] _ in
+                self?.statsUpdateSubject.send()
+            }
+            .store(in: &cancellables)
+
+        trackerDataProvider.trackerDataUpdatesPublisher
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] in
+                self?.refreshTopCompanies()
+            }
+            .store(in: &cancellables)
+
+        refreshTopCompanies()
+    }
+
+    func calculatePrivacyStats() async -> NewTabPagePrivacyStatsClient.PrivacyStatsData {
+        let stats = await privacyStats.fetchPrivacyStats()
+
+        var totalCount: Int64 = 0
+        var otherCount: Int64 = 0
+
+        var companiesStats: [NewTabPagePrivacyStatsClient.TrackerCompany] = stats.compactMap { key, value in
+            totalCount += value
+            guard topCompanies.contains(key) else {
+                otherCount += value
+                return nil
+            }
+            return NewTabPagePrivacyStatsClient.TrackerCompany(count: value, displayName: key)
+        }
+
+        if otherCount > 0 {
+            companiesStats.append(.otherCompanies(count: otherCount))
+        }
+        return NewTabPagePrivacyStatsClient.PrivacyStatsData(totalCount: totalCount, trackerCompanies: companiesStats)
+    }
+
+    private func refreshTopCompanies() {
+        struct TrackerWithPrevalence {
+            let name: String
+            let prevalence: Double
+        }
+
+        let trackers: [TrackerWithPrevalence] = trackerDataProvider.trackerData.entities.values.compactMap { entity in
+            guard let displayName = entity.displayName, let prevalence = entity.prevalence else {
+                return nil
+            }
+            return TrackerWithPrevalence(name: displayName, prevalence: prevalence)
+        }
+
+        let topTrackersArray = trackers.sorted(by: { $0.prevalence > $1.prevalence }).prefix(Const.maxTopCompaniesCount).map(\.name)
+        Logger.privacyStats.debug("top tracker companies: \(topTrackersArray)")
+        topCompanies = Set(topTrackersArray)
+    }
+
+    private enum Const {
+        static let maxTopCompaniesCount: Int = 100
```

Comment:
Minor: Maybe add documentation explaining why 100 is the max here.

Reply:
Sure thing 👍